### PR TITLE
feat(storage): support range-filter scan by sort key

### LINKED ISSUE #589: storage: support range-filter scan by sort key
Our storage engine already supports sorting by sort key. To leverage the property in storage, we can provide a `range_scan` interface on storage, so as to support range filter scan for executors.

To support range_scan, we should be able to start scanning from a sort key in storage engine. This requires two steps:

* record sort key in block index
* support create iterator by sort key
* add interface for executors' use

Currently, the `first_key` field is empty in block builder. We should record `first_key` in order to support seek. Note that https://github.com/risinglightdb/risinglight/issues/588 will add composite sort key support. For this issue, we may only support range filter scan by the first element in composite sort key.

https://github.com/risinglightdb/risinglight/blob/7c5e0d5193708c50b14c98f3efdddfcefbe445cd/src/storage/secondary/block/block_index_builder.rs#L39-L54

After that, we should be able to support create `RowSetIterator` by sort key.

https://github.com/risinglightdb/risinglight/blob/7c5e0d5193708c50b14c98f3efdddfcefbe445cd/src/storage/secondary/rowset/rowset_iterator.rs#L27-L33

The interface should have two new fields `start_key: Option<Vec<u8>>` and `end_key: Option<Vec<u8>>`. RowSetIterator should return rows only in this range. At the same time, the RowSetIterator should directly seek to the block of `start_key` instead of scanning from the beginning.

When everything's ready, we can finally expose range-filter scan interface to the compute layer. This requires us to modify:

https://github.com/risinglightdb/risinglight/blob/7c5e0d5193708c50b14c98f3efdddfcefbe445cd/src/storage/secondary/transaction.rs#L356-L364

### PR DESCRIPTION
**Pull request by Shmiwy**
Signed-off-by: Shmiwy <wyf000219@126.com>
close #589
for now only support range-filter scan by first column and this column should by primary key

### CODE REVIEW

**Review by skyzh — COMMENTED** [review]
Would you please add some unit tests for `RowSetIterator` for range filters before starting review? 🤪

**Comment by Shmiwy** [comment]
> Would you please add some unit tests for `RowSetIterator` for range filters before starting review? 🤪

oh， I forgot again....🥲

File: src/storage/secondary/rowset/disk_rowset.rs
Comment (skyzh) [inline src/storage/secondary/rowset/disk_rowset.rs:128]: Why we only pass `end_sort_key` here? I guess this function should have both `begin_sort_key` and `end_sort_key` as parameter?

File: src/storage/secondary/rowset/disk_rowset.rs
Comment (skyzh) [inline src/storage/secondary/rowset/disk_rowset.rs:176]: This positioning is not accurate. If we have 2 blocks with start of:

```
block 1: a
block 2: d
```

Now we want to seek to `b`. This function will return row_id = 0. However, the first row from block 1 cannot be `b` -- it's `a` instead. Therefore, `RowSetIterator` will still need `begin_sort_key` as parameter, and it will need to filter those keys below start sort key.

File: src/storage/secondary/rowset/rowset_iterator.rs
Comment (skyzh) [inline src/storage/secondary/rowset/rowset_iterator.rs:286]: Directly compare by `end_key < &array.get(len - 1)`?

**Review by skyzh — COMMENTED** [review]
The logic generally correct, but there are some small issues. Comments below.

Also one idea come in my mind: maybe we can make range filter part of the filter condition, so that we don't need to compute which rows should be removed by ourselves, and re-use the current expression framework?

We can doc our RowSetIterator and Scan interface like: begin / end sort key must be included in filter condition as `10 < x and x < 15`, so as to make range filter scan work correctly.

Now we only need to do two additional things compared with the previous code:

1. seek to the block (no need to be accurate, as filter condition will help us filter rows out-of-range)
2. pre-compute the block that contains `end_sort_key` instead of checking the last item in the data chunk. That's because filter condition will help us filter rows out-of-range, and it might incur a lot I/Os before a row can be produced. So if we stop at the block instead of checking every data chunk, things could be easier.

For example, let take this SQL as example:

```
select * from table where x > 10 and x < 15;
```

The RowSetIterator will be created as:

begin_sort_key: [10], end_sort_key: [15], filter: 10 < x and x < 15

Before creating RowSetIterator, we will position the first block with `10` occurrence (namely block A), and the last block with `15` occurrence (namely block B). At the beginning, the RowSetIterator will be positioned at the first row of Block A. When the RowSetIterator goes beyond block B (we can check this by current row id > last row of block B), we return the end of iterator.

Therefore, things could be a lot easier. Also, we can easily support `>` and `>=` conditions.

File: src/storage/secondary/rowset/disk_rowset.rs
Comment (Shmiwy) [reply src/storage/secondary/rowset/disk_rowset.rs:128]: because i put the start row id in seek_pos, and will finish scanning when meet `end_sort_key`

File: src/storage/secondary/rowset/disk_rowset.rs
Comment (Shmiwy) [reply src/storage/secondary/rowset/disk_rowset.rs:176]: oh .the line 166 should be >= instead of >

Comment (skyzh) [reply src/storage/secondary/rowset/disk_rowset.rs:176]: I mean the begin row might not be the same as sort key? In this case, the RowSetIterator might produce more items than expected. You may take a look at the above comment, and put range filters into the filter condition when unit testing.

File: src/storage/secondary/rowset/rowset_iterator.rs
Comment (Shmiwy) [reply src/storage/secondary/rowset/rowset_iterator.rs:286]: Because I found that we implemented  addition, subtraction, multiplication and division between `DataValue`（like int32， int64). so i try reuse the logic to avoid many   forced conversions

File: src/storage/secondary/rowset/disk_rowset.rs
Comment (Shmiwy) [reply src/storage/secondary/rowset/disk_rowset.rs:176]: oh yes. i forget this

File: src/storage/secondary/rowset/rowset_iterator.rs
Comment (Shmiwy) [reply src/storage/secondary/rowset/rowset_iterator.rs:286]: This thing is a little weird. I'll think of something else.😂

**Comment by Shmiwy** [comment]
just fix some bug in this commit, I will make range filter part of the filter condition in the later pr

File: src/storage/secondary/rowset/disk_rowset.rs
Comment (skyzh) [inline src/storage/secondary/rowset/disk_rowset.rs:158]: So we are assuming that the first column of the RowSet is pk, and pk is non-nullable. Should enforce this constraint in binder in later PRs.

File: src/storage/secondary/rowset/disk_rowset.rs
Comment (skyzh) [inline src/storage/secondary/rowset/disk_rowset.rs:180]: One optimization can be done in the future PRs:

* Use `partition_point` function to do binary search, which could be a lot faster.

File: src/storage/secondary/rowset/rowset_iterator.rs
Comment (skyzh) [inline src/storage/secondary/rowset/rowset_iterator.rs:104]: I think we can simply pass `begin_row_id` and `end_row_id` into RowSetIterator? So that we don't need complex logic to compare chunk against sort key in `next_batch`. `begin_row_id` and `end_row_id` include the range to scan, and we use filter scan to filter data. For example, the outer function determines that block 2 - 3 contains data for user specified sort key. So we can pass block 2's first row_id as begin_row_id, and block 4's first row_id - 1 as end_row_id.

File: src/storage/secondary/rowset/rowset_iterator.rs
Comment (skyzh) [inline src/storage/secondary/rowset/rowset_iterator.rs:564]: Filter condition should include the range info. Here we can construct an expression manually -- `180 < InputRef(0) < 195`.

**Review by skyzh — COMMENTED** [review]
Rest LGTM!

**Comment by skyzh** [comment]
> just fix some bug in this commit, I will make range filter part of the filter condition in the later pr

Okay, let me re-do some reviews. If you make range filter part of the filter condition, there are a lot of code that can be removed later :)

File: src/storage/secondary/rowset/rowset_iterator.rs
Comment (skyzh) [reply src/storage/secondary/rowset/rowset_iterator.rs:564]: May fix in later PRs.

File: src/storage/secondary/rowset/rowset_iterator.rs
Comment (skyzh) [reply src/storage/secondary/rowset/rowset_iterator.rs:104]: May fix in later PRs.

File: src/storage/secondary/rowset/rowset_iterator.rs
Comment (skyzh) [inline src/storage/secondary/rowset/rowset_iterator.rs:292]: Can do a binary search. As we will eventually remove this part, looks okay to me now.

**Review by skyzh — APPROVED** [review]
LGTM if you don't want to include range filter in filter condition for now. Let's improve this in future PRs.

**Comment by skyzh** [comment]
So to summarize, there are a list of things to do after this PR gets merged:

* Enforce pk not null, and pks are the first, second, etc., columns in binder. Alternatives: just ensure .iter()'s first column is `StorageColumnRef::Idx(x)`, where x is pk (can be done in optimizer).
* Include range filter in filter condition, and only pass approximate begin/end row id to RowSetIterator. The actual filter will be done by filter condition, and begin/end row id don't need to be accurate.
* Use [`partition_point`](https://doc.rust-lang.org/std/primitive.slice.html#method.partition_point) (or implement binary search by ourselves) when possible.

**Pull request merged**
